src/plonkish/backend/halo2.rs

```diff
@@ -579,32 +635,42 @@ impl Halo2Setup {
     }
 }
 
-pub trait PlonkishHalo2<F: PrimeField> {
-    fn halo2_setup(&mut self, k: u32, rng: BlockRng<DummyRng>) -> Halo2Setup;
+pub trait PlonkishHalo2<F: PrimeField, P> {
```

Comment:
I think this P should have trait, that in the end the only unbounded type is what is passed as witness. This way we could also create provers that take the args in the future.

```
trait Halo2Prover<W> {
    fn generate_proof(&self, W)
}
```

I think I mentioned this, but I don't think it is necessary to pass rng again to generate the proof when it is already passed for the compilation. Is there a reason?

Also, could you answer about the automatic `k` generation? isn't it log2 of nun_rows? perhaps @ed255 knows this.

Reply:
Please check this regarding the K: https://github.com/privacy-scaling-explorations/chiquito/pull/263#issuecomment-2213523912
I think it will be relatively easy to implement as a log2 taking into account the `ConstraintSystem.minimum_rows()`, but this is out of scope of this refactoring as it could be done independently before. I'd like to keep the scope of this PR as small as possible and merge it ASAP.

Reply:
ok, for another task is ok, sorry I didn't see that